### PR TITLE
Losses

### PR DESCRIPTION
**Pull request by AliengirlLiv**
This PR includes the loss function, corresponding to the _train function in the tensorflow code.

Notes for reviewers:
I'd appreciate review on all parts of it, but a few specific points to pay attention to:
1. Whether the loss function uses the correct data.  The samples included have action, prev_action, state, prev_state, and two (different!) reward tensors.  Please double check the right ones are being used in the right places.

1. The `rollout_representation` function (corresponding to `observe` in the TF code).  Ours takes in a prev_action, while theirs doesn't. The two versions seem practically similar, but please double-check me on that.

1. I wasn't able to run this with the main function because of some errors related to image sizes (it seems like parts of the code assumes all images are (3, 64, 64).  Probably not something we need to fix in this PR, but down the line. Also, this PR changes a couple lines in agent.py to get rid of another bug related to state-dimension mismatch.

### CODE REVIEW

**Comment by juliusfrost** [comment]
Looks amazing! I'll check in more after my exam.
